vm_api/src/trace.rs

```diff
@@ -29,26 +35,86 @@ pub struct InvocationTrace {
 //   constructing it.
 #[derive(Clone, Debug)]
 pub struct ExpectInvocation {
-    pub from: Address,
+    pub from: ActorID,
     pub to: Address,
     pub method: MethodNum,
     pub value: Option<TokenAmount>,
     pub params: Option<Option<IpldBlock>>,
-    pub code: ExitCode,
-    pub ret: Option<Option<IpldBlock>>,
+    pub result: ExpectResult,
     pub subinvocs: Option<Vec<ExpectInvocation>>,
 }
 
+#[derive(Clone, Debug)]
+pub enum ExpectResult {
+    /// Separate from CallReturn so that matching on return_value is optional
+    ExpectReturn {
+        return_value: Option<Option<IpldBlock>>,
+        exit_code: ExitCode,
+    },
+    ExpectError {
+        reason: String,
+        errno: ErrorNumber,
+    },
+}
+
+impl ExpectResult {
+    pub fn ok(return_value: Option<Option<IpldBlock>>) -> ExpectResult {
+        Self::ExpectReturn { return_value, exit_code: ExitCode::OK }
+    }
+}
+
```

Comment:
This is nice but I think ended up more code than it's worth. I suggest just flattening the exit code, return value, and `Option<ErrorNumber>` (skip the string) into both InvocationTrace and ExpectInvocation. Use some sentinel exit code when there's an error, and check for error number first when matching, checking the exit code only if no error was expected.